src/cnav_msg.c

```diff
@@ -0,0 +1,515 @@
+    }
+    else
+    {
+      /* No preamble or preamble and less than 300 bits decoded */
+    }
+  }
+}
+
+/**
+ * Invert message bits in the buffer.
+ *
+ * The method inverts bits of the decoded data.
+ *
+ * \param[in,out] part Decoder component with a message buffer.
+ *
+ * \return None
+ */
+void _cnav_msg_invert(cnav_v27_part_t *part)
+{
+  for (size_t i = 0; i < sizeof(part->decoded); i++) {
+    part->decoded[i] ^= 0xFFu;
+  }
+}
+
+/**
+ * Performs CNAV message decoding.
+ *
+ * This function decoded CNAV message, if the following conditions are met:
+ * - Buffer contains 300 bits.
+ * - First 8 bits are matching direct or inverse preamble.
+ * - Message data CRC matches one in the buffer.
+ *
+ * In case the message starts with inverted preamble, the data is inverted
+ * before parsing.
+ *
+ * \param[in,out] part Decoder component.
+ * \param[out]    msg  Container for a decoded message.
+ * \param[out]    delay Delay of the message in symbols.
+ *
+ * \retval true The message has been decoded, and \a msg container is populated.
+ * \retval false Not enough data or CRC is not correct.
+ *
+ * \private
+ */
+bool _cnav_msg_decode(cnav_v27_part_t *part, cnav_msg_t *msg, u32 *delay)
+{
+  bool res = false;
+  if (GPS_CNAV_MSG_LENGTH <= part->n_decoded) {
+    if (part->crc_ok) {
+      /* CRC is OK */
+
+      if (part->invert) {
+        _cnav_msg_invert(part);
+      }
+
+      msg->prn    = getbitu(part->decoded, 8, 6);
+      msg->msg_id = getbitu(part->decoded, 14, 6);
+      msg->tow    = getbitu(part->decoded, 20, 17);
+      msg->alert  = getbitu(part->decoded, 37, 1) ? true : false;
+
+      *delay = (part->n_decoded - GPS_CNAV_MSG_LENGTH + GPS_L2C_V27_DELAY_BITS)
+               * 2 + part->n_symbols;
+
+      if (part->invert) {
+        _cnav_msg_invert(part);
+      }
+      res = true;
+    } else {
+      /* CRC mismatch - no decoding */
+    }
+    _cnav_bitlshift(part->decoded, sizeof(part->decoded), GPS_CNAV_MSG_LENGTH);
+    part->n_decoded -= GPS_CNAV_MSG_LENGTH;
+  }
+
+  return res;
+}
+
+/**
+ * Initialize CNAV decoder.
+ *
+ * CNAV decoder contains two Viterbi decoders that are used to estimate bit and
+ * message boundary.
+ *
+ * \param[out] dec Decoder structure.
+ *
+ * \return None
+ */
+void cnav_msg_decoder_init(cnav_msg_decoder_t *dec)
+{
+  memset(dec, 0, sizeof(*dec));
+  v27_init(&dec->part1.dec,
+           dec->part1.decisions,
+           GPS_L2_V27_HISTORY_LENGTH_BITS,
+           cnav_msg_decoder_get_poly(),
+           0);
+  v27_init(&dec->part2.dec,
+           dec->part2.decisions,
+           GPS_L2_V27_HISTORY_LENGTH_BITS,
+           cnav_msg_decoder_get_poly(),
+           0);
+  dec->part1.init = true;
+  dec->part2.init = true;
+  _cnav_add_symbol(&dec->part2, 0x80);
+}
+
+/**
+ * Adds a received symbol to decoder.
+ *
+ * \param[in,out] dec    Decoder object.
+ * \param[in]     symbol Symbol value probability, where 0x00 - 100% of 0,
+ *                       0xFF - 100% of 1.
```

Comment:
Might want to double check, I thought the `symbol` was passed as a signed number, i.e. 2's complement